[JIT] Add special cases batch_norm, instance_norm and dropout for SchemaInfo

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81518
* #81444
* #81439
* #81352
* __->__ #81007

- Added special cases for detach in is_non_deterministic() check and batch_norm and instance_norm in is_mutable() check in SchemaInfo().
- Added tests for the above special cases for detach, batch_norm and instance_norm.

Commit:
[JIT] Add special cases batch_norm, instance_norm and dropout for SchemaInfo

[ghstack-poisoned]

Commit:
Update on "[JIT] Add special cases batch_norm, instance_norm and dropout for SchemaInfo"




[ghstack-poisoned]

davidberard98 commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -67,7 +67,26 @@ bool SchemaInfo::is_mutable(size_t index) {
       input_alias_map_[index].end(),
-      [this](size_t index) { return this->schema_.is_mutable(index); });
+      [this](size_t index) {
+        static const char* batch_norm_literal =
```
A few suggestions here:

1. Can you do the schema parsing statically so that it doesn't have to happen each time?
2. Can you move the special case behavior outside of the `std::any_of`?  The special case behavior doesn't really have any dependence on index, it would probably be easier to understand just as `if (this_schema == batch_norm_literal && (has_running_mean || has_running_var) { return training_is_true; }`

goldenxuett replied:
1. Yeah I'll clean that up and also make the special case look more readable.
2. Technically there actually is a dependence on index. For instance, with batch_norm if the input tensor value is aliasing the running_mean and training = true, then the input tensor will now be mutable which is found through std::any_of

davidberard98 reviewed (commented):
Nice, it's good to see the special cases in here! Just had some comments about how to handle the special cases in SchemaInfo

Commit:
Update on "[JIT] Add special cases batch_norm, instance_norm and dropout for SchemaInfo"


- Added special cases for detach in is_non_deterministic() check and batch_norm and instance_norm in is_mutable() check in SchemaInfo().
- Added tests for the above special cases for detach, batch_norm and instance_norm.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add special cases batch_norm, instance_norm and dropout for SchemaInfo"


- Added special cases for detach in is_non_deterministic() check and batch_norm and instance_norm in is_mutable() check in SchemaInfo().
- Added tests for the above special cases for detach, batch_norm and instance_norm.

[ghstack-poisoned]

davidberard98 commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -67,10 +67,34 @@ bool SchemaInfo::is_mutable(size_t index) {
       input_alias_map_[index].begin(),
       input_alias_map_[index].end(),
-      [this](size_t index) { return this->schema_.is_mutable(index); });
+      [this](size_t index) {
```
sorry, one more request I didn't notice last time - could you rename this from `index` to something that doesn't shadow a previous variable? (actually I got confused in the last review and that's why I had the previous suggestion).

Commit:
Update on "[JIT] Add special cases batch_norm, instance_norm and dropout for SchemaInfo"


- Added special cases for detach in is_non_deterministic() check and batch_norm and instance_norm in is_mutable() check in SchemaInfo().
- Added tests for the above special cases for detach, batch_norm and instance_norm.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add special cases batch_norm, instance_norm and dropout for SchemaInfo"


- Added special cases for detach in is_non_deterministic() check and batch_norm and instance_norm in is_mutable() check in SchemaInfo().
- Added tests for the above special cases for detach, batch_norm and instance_norm.

[ghstack-poisoned]

davidberard98 reviewed (approved):
nice, looks good!

Commit:
Update on "[JIT] Add special cases batch_norm, instance_norm and dropout for SchemaInfo"


- Added special cases for detach in is_non_deterministic() check and batch_norm and instance_norm in is_mutable() check in SchemaInfo().
- Added tests for the above special cases for detach, batch_norm and instance_norm.

[ghstack-poisoned]

goldenxuett commented:
@pytorchbot merge
